File: java/19-RemoveNthNodeFromEndOfList.java
/*

Given a linked list, remove the nth node from the end of list and return its head.

For example,

   Given linked list: 1->2->3->4->5, and n = 2.

   After removing the second node from the end, the linked list becomes 1->2->3->5.
Note:
Given n will always be valid.
Try to do this in one pass.

*/


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy;
        for(int i = 0; i<n; i++) {
            fast = fast.next; // we know fast always exist
        }
        ListNode slow = head;
        ListNode prev = dummy;
        while(fast.next!=null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next;
        }
        prev.next = slow.next;
        return dummy.next;
    }
}

class Solution1 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        int k = n;
        ListNode curr = head;
        n = 0;
        while(curr != null) {
            curr = curr.next;
            ++n;
        }
        n = n-k+1;
        
        curr = head;
        ListNode prev = dummy;
        for(int i = 0; i<n-1; i++) {
            prev = curr;
            curr = curr.next;
        }
        prev.next = curr.next;
        return dummy.next;
    }
}

File: c++/19-RemoveNthNodeFromEndOfList.cpp
/*

Given a linked list, remove the nth node from the end of list and return its head.

For example,

   Given linked list: 1->2->3->4->5, and n = 2.

   After removing the second node from the end, the linked list becomes 1->2->3->5.
Note:
Given n will always be valid.
Try to do this in one pass.

*/


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */


class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int k) {
        if(!head) return head;
        int n = 0;
        ListNode *curr = head;
        while(curr) {++n; curr= curr->next;}
        k = n-k+1;
        int cnt = 0;
        ListNode dummy{0};
        curr = &dummy;
        curr->next = head;
        ListNode *prev = NULL;
        while(cnt !=k ) {
            prev = curr;
            curr = curr->next;
            ++cnt;
        }
        prev->next = curr->next;
        delete curr;
        return dummy.next;
    }
};

class Solution1 {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
            ListNode *slow = head, *fast = head, *pre = NULL;

            while(n > 0) {
                fast = fast->next;
                --n;
            }

            while(fast) {
                pre = slow;
                slow = slow->next;
                fast = fast->next;
            }

            if(!pre && !slow->next)
                return NULL;

            if(!pre && slow->next)
                return slow->next;

            pre->next = slow->next;
            delete slow;

            return head;        
    }
};
